crypto/threads_pthread.c

```diff
@@ -42,12 +44,483 @@
+}
+
+static void retire_qp(struct rcu_lock_internal *ilock,
+                      struct rcu_qp *qp)
+{
+    pthread_mutex_lock(&ilock->alloc_lock);
+    ilock->writers_alloced--;
+    pthread_cond_signal(&ilock->alloc_signal);
+    pthread_mutex_unlock(&ilock->alloc_lock);
+}
+
+static struct rcu_qp *allocate_new_qp_group(struct rcu_lock_internal *ilock,
+                                            int count)
+{
+    struct rcu_qp *new =
+        OPENSSL_zalloc(sizeof(struct rcu_qp) * count);
+
+    ilock->group_count = count;
+    return new;
+}
+
+/* Public WRITE api */
+void ossl_rcu_write_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_lock(&ilock->write_lock);
+}
+
+void ossl_rcu_write_unlock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_unlock(&ilock->write_lock);
+}
+
+/* Public WRITE api for synchronization */
+void ossl_synchronize_rcu(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_qp *qp;
+    struct rcu_lock_internal *ilock;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /* before we do anything else, lets grab the cb list */
+    ilock = (struct rcu_lock_internal *)lock;
+    cb_items = __ATOMIC_EXCHANGE_N(&ilock->cb_items, NULL, __ATOMIC_SEQ_CST);
+
+    qp = update_qp(ilock);
+
+    /* wait for the reader count to reach zero */
+    do {
+        count = (uint64_t)__ATOMIC_LOAD_N(&qp->users, __ATOMIC_SEQ_CST);
+    } while (READER_COUNT(count) != 0);
+
+    /* retire in order */
+    pthread_mutex_lock(&ilock->prior_lock);
+    while (ilock->next_to_retire != ID_VAL(count))
+        pthread_cond_wait(&ilock->prior_signal, &ilock->prior_lock);
+    ilock->next_to_retire++;
+    pthread_cond_broadcast(&ilock->prior_signal);
+    pthread_mutex_unlock(&ilock->prior_lock);
+
+    retire_qp(ilock, qp);
+
+    /* handle any callbacks that we have */
+    while (cb_items != NULL) {
+        tmpcb = cb_items;
+        cb_items = cb_items->next;
+        tmpcb->fn(tmpcb->data);
+        CRYPTO_free(tmpcb, NULL, 0);
+    }
+
+    /* and we're done */
+    return;
+}
+
+/*
+ * Public WRITE api to schedule a callback after a
+ * completed synchronization
+ */
+void ossl_rcu_call(CRYPTO_RCU_LOCK *lock, rcu_cb_fn cb, void *data)
+{
+    struct rcu_cb_item *new =
+        OPENSSL_zalloc(sizeof(struct rcu_cb_item));
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    new->data = data;
+    new->fn = cb;
+    new->next = __ATOMIC_EXCHANGE_N(&ilock->cb_items, new, __ATOMIC_SEQ_CST);
+}
+
+void *ossl_rcu_uptr_derefrence(void **p)
+{
+    return (void *)__ATOMIC_LOAD_N(p, __ATOMIC_ACQUIRE);
+}
+
+void ossl_rcu_assign_uptr(void **p, void **v)
+{
+    __ATOMIC_STORE(p, v, __ATOMIC_RELEASE);
+}
+
+static CRYPTO_ONCE rcu_init_once = CRYPTO_ONCE_STATIC_INIT;
+
+CRYPTO_RCU_LOCK *ossl_rcu_lock_new(int num_writers)
```

Comment:
This isn't actually the number of writers - but the group size. A group size of 2 means you can have 1 writer. So I'd suggest defaulting this to 1 and passing `num_writers + 1` to the `allocate_new_qp_group` call. It really does make sense for you to pass in the number of writers that you want, since the group size won't mean much to callers of this function.